replacing __Verify_noErr on macOS, generic implementation

csomor opened the pull request:
macros for verifying methods don’t fail with native error codes

csomor committed:
Add wxVERIFY_NOERR_XXX macros

These macros provide portable equivalent for __Verify_noErr from macOS.

They are different from asserts, as the expression given to the macro is
always evaluated and then, if the result is not zero, and asserts are
on, an assert containing a human readable message with the description
of the error code is triggered.

Closes https://github.com/wxWidgets/wxWidgets/pull/1973

Co-authored-by: Vadim Zeitlin <vadim@wxwidgets.org>

vadz commented:
Looks good, thanks, but we should document the new macros if they're meant to be public.

And, as you probably already noticed, this breaks wxiOS build, hopefully fixing this will be simple for you...

csomor commented:
@vadz that do you think about this one, if possible, I'd like to have it in 3.1.4 as well

vadz commented:
OK, I'll add the docs and merge.

vadz commented on include/wx/debug.h:
```diff
@@ -328,6 +335,10 @@ extern WXDLLIMPEXP_BASE void wxOnAssert(const char *file,
 
     #define wxASSERT(cond)
     #define wxASSERT_MSG(cond, msg)
+
+    #define wxASSERT_NOERR_MSG(errorCode, msg)
+    #define wxASSERT_NOERR(errorCode)
+
```
Just FYI: I'm removing those as this looks like a leftover: these macros have no non-trivial definitions anywhere.

csomor replied:
yes, sorry, leftovers, I first wanted to use asserts, but then I realized, that they are in some parts used directly with the OS call itself, so CHECK was more appropriate

vadz commented:
Sorry, there seem to be some more serious problems here:

- `wxOnAssertWithErrorCode()` is only defined inside `wxUSE_UNICODE` but used from `wxVERIFY_NOERR_MSG_AT` which doesn't have any such guard, so this almost certainly breaks the build with `wxUSE_UNICODE==0`. I know it's not very important, but as long as we have `wxUSE_UNICODE` at all, we shouldn't break building without it...
- Worse, `wxVERIFY_NOERR_MSG_AT` is also defined outside of any `wxDEBUG_LEVEL` checks, so this breaks the build with `wxDEBUG_LEVEL==0` and we do want to support this.

I may try fixing this if it's really, really critical, but perhaps we could live it for after 3.1.4? The potential for breakage is rather high here... Please let me know if you absolutely need this in 3.1.4.

csomor commented:
> Sorry, there seem to be some more serious problems here:
> 
> * `wxOnAssertWithErrorCode()` is only defined inside `wxUSE_UNICODE` but used from `wxVERIFY_NOERR_MSG_AT` which doesn't have any such guard, so this almost certainly breaks the build with `wxUSE_UNICODE==0`. I know it's not very important, but as long as we have `wxUSE_UNICODE` at all, we shouldn't break building without it...
> * Worse, `wxVERIFY_NOERR_MSG_AT` is also defined outside of any `wxDEBUG_LEVEL` checks, so this breaks the build with `wxDEBUG_LEVEL==0` and we do want to support this.
> 
> I may try fixing this if it's really, really critical, but perhaps we could live it for after 3.1.4? The potential for breakage is rather high here... Please let me know if you absolutely need this in 3.1.4.

no not critical, don't waste time, I was aware of the unicode requirement, but I'm not sure I understand the other implication correctly

the way things are used with __Verify_noErr, at some places directly the OS call is inside the parenthesis, so this part has to be executed always, no matter which flags are set or not

vadz commented:
Yes, I understand that `wxVERIFY` is like `wxCHECK` in this respect, but this means that it has to use something like `wxFAIL_COND_MSG`, i.e. a macro which is still defined (as nothing) even when debugging checks are completely turned off.

vadz commented:
I've force pushed the commit I wanted to apply before I realized there were more problems with it.

csomor commented:
> I've force pushed the commit I wanted to apply before I realized there were more problems with it.

I've added another commit in direction of the solution, but this does not show up here somehow ? I've put the `wxVERIFY_NOERR_MSG_AT` as it is in a #if part and a minimal implementation in the #else:

```
#if wxDEBUG_LEVEL && wxUSE_UNICODE
(...)
#else 
#define wxVERIFY_NOERR_MSG_AT(cond, msg, file, line, func)                \
wxSTATEMENT_MACRO_BEGIN                                                   \
    cond;                                                                 \
wxSTATEMENT_MACRO_END
#endif
```

vadz commented:
Sorry, I didn't realize you were working on this, I must have overwritten your changes with my force push :-( Could you please try rebasing your local branch on this one and pushing again?

BTW, I wonder if we really need a separate `wxOnAssertWithErrorCode()`. Couldn't we just use the existing `wxOnAssert()` with `wxSysErrorMsgStr()`?

csomor committed:
fixing non-unicode and wxDEBUG_LEVEL == 0 builds

csomor commented:
> Sorry, I didn't realize you were working on this, I must have overwritten your changes with my force push :-( Could you please try rebasing your local branch on this one and pushing again?

done
 
> BTW, I wonder if we really need a separate `wxOnAssertWithErrorCode()`. Couldn't we just use the existing `wxOnAssert()` with `wxSysErrorMsgStr()`?

I thought about that, but I thought I'd better keep the macro as small as possible, but if you don't think I should care about that, then I can do that of course

StefanBruens commented:
@csomor - are you still working on this?
@vadz - the macOS changes are, based on the Milestone open tasks, the only things currently blocking a 3.1.5 release. Is there anything else? I really would like to have a release which can be used for Phoenix. Trying to build e.g. KiCad, you either end up with a quite old wxWidgets/Phoenix version, or have to bundle a heavily patched version or some (almost random) git tag.

csomor committed:
Simplifying implementation

Getting rid of more sophisticated options, base on already existing wxFAIL_COND_MSG

csomor committed:
Removing white space change

csomor commented:
> @csomor - are you still working on this?

I've committed the changes using the direction @vadz hinted at

vadz commented on include/wx/debug.h:
```diff
@@ -390,6 +390,32 @@ extern void WXDLLIMPEXP_BASE wxAbort();
+
+// This macro checks if the evaluation of cond, having a return value of
+// OS Error type, is zero, ie no error occurred, and calls the assert handler
+// with the provided message if it isn't and finally traps if the special flag
+// indicating that it should do it was set by the handler.
```
The part about the special flag is not true any longer.

vadz commented on include/wx/debug.h:
```diff
@@ -390,6 +390,32 @@ extern void WXDLLIMPEXP_BASE wxAbort();
+        }                                                                 \
+        else                                                              \
+        {                                                                 \
+            wxFAIL_COND_MSG(#cond, wxSysErrorMsgStr(cond));               \
```
Should use `evalOnceErrorCode` instead of `code` here.

vadz commented on src/common/log.cpp:
```diff
@@ -1121,6 +1125,12 @@ static const wxChar* GetSysErrorMsg(wxChar* szBuf, size_t sizeBuf, unsigned long
+#elif defined(__WXOSX_COCOA__)
+        int iErrorCode = (int)nErrCode;
+        if ( iErrorCode < 0 )
+            strncpy(buffer,GetMacOSStatusErrorString(iErrorCode), sizeof(buffer));
```
This function is deprecated since 10.8, is there really no replacement (something in `NSError`?) that we could use instead?

csomor replied:
unfortunately not ... at least not something that is as precise

https://stackoverflow.com/questions/14699604/replacements-for-getmacosstatuserrorstring-getmacosstatuscommentstring

and the fact that the error strings are not useful is still true in 10.15 ...

vadz commented on include/wx/debug.h:
```diff
@@ -390,6 +390,32 @@ extern void WXDLLIMPEXP_BASE wxAbort();
+        }                                                                 \
+    wxSTATEMENT_MACRO_END
+
+#else
+
+#define wxVERIFY_NOERR(cond)                                              \
+    wxSTATEMENT_MACRO_BEGIN                                               \
+        cond;                                                             \
+    wxSTATEMENT_MACRO_END
+
+#endif
```
I don't think we need this part at all, actually, `wxFAIL_COND_MSG` already expands to nothing when debug checks are disabled, so I believe we should just always use the definition in `#if`. Am I missing something here?

csomor replied:
You are absolutely right, it was a left over from previous attempts, where I used my own macros, but now it's not needed anymore

vadz reviewed (commented):
Sorry for taking so long to get back to this!

I can apply after addressing the comments I've left if you don't mind, but please let me know if you know of any better replacement for `GetMacOSStatusErrorString()`. TIA!

vadz commented:
Stefan, I'm finally merging this but with the new macro in `include/wx/osx/core/private.h` rather than in the public header. We can always make it public later, but for now keeping it private is simpler as we don't have to decided whether it's the best name for it (IMO it should somehow indicate that it checks for the _system_ error), whether there should be a separate `_MSG` variant etc, and we also don't need to document it.

I hope you don't mind this but, again, if you do, we can always make it public later.
